File: C++/Leetcode/SwapNodesInLinkedList.cpp
// success
// https://leetcode.com/problems/swapping-nodes-in-a-linked-list/
#include <iostream>
#include <string>
#include <math.h>
#include <iomanip>
#include <list>
#include <vector>
#include <stack>
#include <queue>
#include <map>
#include <set>
#include <algorithm>
#include <numeric>
#include <climits>

using namespace std;

// Definition for singly-linked list.
struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr){};
    ListNode(int x) : val(x), next(nullptr){};
    ListNode(int x, ListNode *next) : val(x), next(next){};
};

class Solution
{
public:
    ListNode *swapNodes(ListNode *head, int k)
    {
        if (!head->next)
            return head;
        ListNode *first = head;
        ListNode *slow = head;
        ListNode *fast = head;
        for (int i = 1; i < k; i++)
        {
            first = first->next;
            fast = fast->next;
        }
        while (fast->next)
        {
            slow = slow->next;
            fast = fast->next;
        }
        int temp = first->val;
        first->val = slow->val;
        slow->val = temp;
        return head;
    }
};

void insert(ListNode *&head, int x)
{
    if (head == nullptr)
    {
        head = new ListNode(x);
        return;
    }
    ListNode *current = head;
    while (current->next != nullptr)
        current = current->next;
    current->next = new ListNode(x);
}

void printList(ListNode *head)
{
    while (head != nullptr)
    {
        cout << head->val << " ";
        head = head->next;
    }
    cout << '\n';
}

int main()
{
    Solution s;
    ListNode *head = new ListNode(1);
    insert(head, 2);
    insert(head, 3);
    insert(head, 4);
    insert(head, 5);
    printList(head);
    head = s.swapNodes(head, 1);
    printList(head);
    return 0;
}

File: Java/com/leetcode/SwapNodesInLinkedList.java
// success
// https://leetcode.com/problems/swapping-nodes-in-a-linked-list/
package com.leetcode;


// Definition for singly-linked list.
class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

class SwapNodesInLinkedList {
    public ListNode swapNodes(ListNode head, int k) {
        ListNode slow = head;
        ListNode fast = head;

        for (int i = 0; i < k - 1; i++) {
            fast = fast.next;
        }
        
        ListNode first = fast;

        while(fast.next != null) {
            slow = slow.next;
            fast = fast.next;
        }

        ListNode second = slow;

        int temp = first.val;
        first.val = second.val;
        second.val = temp;

        return head;
    }

    public static void main(String[] args) {
        SwapNodesInLinkedList s = new SwapNodesInLinkedList();
        ListNode l = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));

	ListNode l1 = s.swapNodes(l, 2);
	for (int i = 0; i < 5; i++) {
	    System.out.print(l1.val + " ");
	    l1 = l1.next;
	}
    }
}